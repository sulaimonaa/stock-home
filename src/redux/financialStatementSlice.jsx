import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const API_KEY = '2a983b252a14dce3b861a39db579c57f';

export const FetchFinancialStatements = createAsyncThunk(
  'financialstatements/fetchstatement',
  async () => {
    try {
      const response = await fetch(
        `https://financialmodelingprep.com/api/v3/financial-statement-symbol-lists?apikey=${API_KEY}`
      );
      if (!response.ok) {
        throw new Error('Network response was not ok');
      }
      const data = await response.json();
      return data;
    } catch (error) {
      throw new Error('Something went wrong with fetching data');
    }
  }
);

const initialState = {
  value: [],
  status: 'idle',
  error: false,
};

const financialStatementReducer = createSlice({
  name: 'financialstatements',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(FetchFinancialStatements.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(FetchFinancialStatements.fulfilled, (state, action) => ({
        ...state,
        status: 'successful',
        value: action.payload,
      }))
      .addCase(FetchFinancialStatements.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: action.error.message,
      }))
  },
});

export default financialStatementReducer.reducer;
