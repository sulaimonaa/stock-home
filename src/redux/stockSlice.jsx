import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const API_KEY = '2a983b252a14dce3b861a39db579c57f';

export const FetchStockLists = createAsyncThunk(
  'stocks/fetchstocks',
  async () => {
    try {
      const response = await fetch(
        `https://financialmodelingprep.com/api/v3/stock/list?apikey=${API_KEY}`
      );
      if (!response.ok) {
        throw new Error('Network response was not ok');
      }
      const data = await response.json();
      return data;
    } catch (error) {
      throw new Error('Something went wrong with fetching data');
    }
  }
);

const initialState = {
  value: [],
  status: 'idle',
  error: false,
};

const stockListReducer = createSlice({
  name: 'stocks',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(FetchStockLists.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(FetchStockLists.fulfilled, (state, action) => ({
        ...state,
        status: 'successful',
        value: action.payload,
      }))
      .addCase(FetchStockLists.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: action.error.message,
      }))
  },
});

export default stockListReducer.reducer;
