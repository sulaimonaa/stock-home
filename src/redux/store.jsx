import { configureStore, getDefaultMiddleware } from "@reduxjs/toolkit";
import thunk from "redux-thunk";
import financialStatementReducer from './financialStatementSlice';
import stockListReducer from './stockSlice'

const store = configureStore({
    reducer: {
        financialstatements: financialStatementReducer,
        stocks: stockListReducer
    },
    middleware: [thunk, ...getDefaultMiddleware()]
});

export default store;
