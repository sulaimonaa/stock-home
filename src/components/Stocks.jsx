import React, { useEffect } from "react"; // eslint-disable-line no-undef, no-unused-vars
import { useDispatch, useSelector } from "react-redux";
import { FetchStockLists } from "../redux/stockSlice";
import DisplayStocks from "./DisplayStocks";
import Navigation from './OtherNav';
import '../styles/stocks.css';

const Stocks = () => {
    const stockList = useSelector((state) => state.stocks.value);
    const limitedData = stockList.slice(0, 150);
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(FetchStockLists());
    }, [dispatch])

    return(
        <>
        <Navigation />
        <div className="stockWrap">
            <h3>Stock Market</h3>
            <table className="table">
                <thead>
                    <tr>
                        <th>Symbol</th>
                        <th>Name</th>
                        <th>Exchange</th>
                        <th>Price</th>
                    </tr>
                </thead>
                <tbody>
                    {limitedData.map((item, index) =>(
                        <DisplayStocks 
                          key={index}
                          sSymbol={item.symbol}
                          sName={item.name}
                          sPrice={item.price}
                          sExchange={item.exchange}
                        />
                    ))}
                </tbody>
            </table>
        </div>
        </>
    )
}
export default Stocks;