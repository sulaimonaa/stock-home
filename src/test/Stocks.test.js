import React from 'react'; // eslint-disable-line no-undef, no-unused-vars
import { render, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Stocks from '../components/Stocks';
import { FetchStockLists } from '../redux/stockSlice';
import { BrowserRouter } from 'react-router-dom'; // Import BrowserRouter

/* eslint-disable no-undef, no-unused-vars */
jest.mock('../redux/stockSlice', () => ({
  FetchStockLists: jest.fn(),
}));
/* eslint-disable no-undef, no-unused-vars */
describe('Stocks', () => {
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);

  it('fetches stock lists and displays them', async () => {
    const stockList = [
      { symbol: 'AAPL', name: 'Apple Inc.', price: 150.00, exchange: 'NASDAQ' },
      // Add more mock stock data as needed
    ];

    // Mock the FetchStockLists action creator
    FetchStockLists.mockReturnValue(() => ({ type: 'stockSlice/FetchStockLists' }));

    const store = mockStore({
      stocks: { value: stockList },
    });

    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Stocks />
        </BrowserRouter>
      </Provider>
    );

    // Wait for component effects to run (e.g., fetch)
    await waitFor(() => {
      expect(FetchStockLists).toHaveBeenCalled();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
