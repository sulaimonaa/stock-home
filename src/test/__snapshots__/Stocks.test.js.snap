// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stocks fetches stock lists and displays them 1`] = `
<DocumentFragment>
  <div
    class="nav"
  >
    <div
      class="container navbar"
    >
      <div
        class="home-icon"
      >
        <img
          alt="previous icon"
          src="mocked-file-url"
        />
      </div>
      <div
        class="home-icon"
      >
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          <img
            alt="home icon"
            src="mocked-file-url"
          />
        </a>
      </div>
      <div
        class="home-icon"
      >
        <a
          class=""
          href="/stocks"
        >
          <img
            alt="home icon"
            src="mocked-file-url"
          />
        </a>
      </div>
      <div
        class="menu-icon"
      >
        <a
          class=""
          href="/about"
        >
          <img
            alt="menu icon"
            src="mocked-file-url"
          />
        </a>
      </div>
    </div>
  </div>
  <div
    class="stockWrap"
  >
    <h3>
      Stock Market
    </h3>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Symbol
          </th>
          <th>
            Name
          </th>
          <th>
            Exchange
          </th>
          <th>
            Price
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Apple Inc.
          </td>
          <td>
            AAPL
          </td>
          <td>
            NASDAQ
          </td>
          <td>
            150
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
